### PR TITLE
Refactoring the code for detecting the project type

### PR DESCRIPTION
**Pull request by saketjajoo**
Currently, the `detect_project_type` parameter `src_dir` is a bit ambiguous. While checking for "docker" project type, the parameter is assumed to be a file name (or a complete file path), but while checking for other project types, it is assumed to be a path to a directory (and not a file).

This PR attempts to fix this ambiguity as well as refactor the code a bit.

### CODE REVIEW

File: depscan/lib/utils.py [hunk 1]
```diff
@@ -95,90 +70,42 @@ def is_exe(src):
...
-        project_types.append("haskell")
-    if find_files(src_dir, "mix.lock", quick=True):
-        project_types.append("elixir")
-    if find_files(
-        os.path.join(src_dir, ".github", "workflows"),
-        ".yml",
-        quick=True,
-        filter=False,
-    ):
-        project_types.append("github")
```
Comment (prabhu) [inline depscan/lib/utils.py:170]: Looks like github got missed out

Comment (saketjajoo) [reply depscan/lib/utils.py:170]: It's on line 91.

File: depscan/lib/utils.py [hunk 2]
```diff
@@ -95,90 +70,42 @@ def is_exe(src):
...
-        project_types.append("jenkins")
-    if find_files(src_dir, ".yml", quick=True) or find_files(
-        src_dir, ".yaml", quick=True
-    ):
-        project_types.append("yaml-manifest")
-    return project_types
+    for root, dirs, files in os.walk(src_dir):
+        filter_ignored_dirs(dirs)
+        for file in files:
+            project_types.extend(detect_project_type_from_file(os.path.abspath(os.path.join(root, file))))
```
Comment (prabhu) [inline depscan/lib/utils.py:107]: Can we add a quick mode so we don't check every file?

Comment (cerrussell) [reply depscan/lib/utils.py:107]: @saketjajoo Is this feasible?

Comment (saketjajoo) [reply depscan/lib/utils.py:107]: Theoretically, yes this is feasible. I can just tweak the original implementation to include everything, but the code readability would get affected (as I would have to rewrite the same logic twice - one for filename as input and one for directory as input). However, in this approach, the code is a bit more readable and also more accurate but a bit slow as compared to the original one. Hence there's a tradeoff between (code readability + correctness) vs performance.

Comment (prabhu) [reply depscan/lib/utils.py:107]: @saketjajoo @cerrussell Shall we explore an alternative where depscan doesn't attempt any detection at all but simply invokes cdxgen with either `-t universal` or without any type when `--no-universal` is passed? This way we reuse the type detection performed by cdxgen.

Comment (cerrussell) [reply depscan/lib/utils.py:107]: @prabhu Does that mean we couldn't use --bom?

Comment (prabhu) [reply depscan/lib/utils.py:107]: When invoked with --bom, can we parse the bom document to collect the types?

**Comment by saketjajoo** [comment]
This is continued in the discussions: https://github.com/owasp-dep-scan/dep-scan/discussions/139.

**Pull request closed without merging**